lxd/device/cdi/spec.go

```diff
@@ -0,0 +1,104 @@
+package cdi
+
+import (
+	"fmt"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/NVIDIA/nvidia-container-toolkit/pkg/nvcdi"
+	cdiSpec "tags.cncf.io/container-device-interface/pkg/cdi"
+	"tags.cncf.io/container-device-interface/specs-go"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+const (
+	// defaultNvidiaTegraMountSpecPath is default location of CSV files that define the modifications required to the OCI spec.
+	defaultNvidiaTegraMountSpecPath = "/etc/nvidia-container-runtime/host-files-for-container.d"
+)
+
+// DefaultNvidiaTegraCSVFiles returns the default CSV files for the Nvidia Tegra platform.
+func DefaultNvidiaTegraCSVFiles(rootPath string) []string {
+	files := []string{
+		"devices.csv",
+		"drivers.csv",
+		"l4t.csv",
+	}
+
+	var paths []string
+	for _, file := range files {
+		paths = append(paths, filepath.Join(rootPath, defaultNvidiaTegraMountSpecPath, file))
+	}
+
+	return paths
+}
+
+// generateNvidiaSpec generates a CDI spec for an Nvidia vendor.
+func generateNvidiaSpec(cdiID ID) (*specs.Spec, error) {
+	mode := nvcdi.ModeAuto
+	if cdiID.Class == IGPU {
+		mode = nvcdi.ModeCSV
+	}
+
+	deviceNamer, err := nvcdi.NewDeviceNamer(nvcdi.DeviceNameStrategyIndex)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create device namer: %v", err)
+	}
+
+	deviceNamers := []nvcdi.DeviceNamer{deviceNamer}
+	nvidiaCTKPath, err := exec.LookPath("nvidia-ctk")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to find the nvidia-ctk binary: %v", err)
+	}
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	cdilib, err := nvcdi.New(
+		nvcdi.WithLogger(&CDILogger{}),
+		nvcdi.WithDriverRoot(rootPath),
+		nvcdi.WithDevRoot(rootPath),
+		nvcdi.WithNVIDIACTKPath(nvidiaCTKPath),
+		nvcdi.WithDeviceNamers(deviceNamers...),
+		nvcdi.WithMode(mode),
+		nvcdi.WithCSVFiles(DefaultNvidiaTegraCSVFiles(rootPath)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create CDI library: %v", err)
+	}
+
+	deviceSpecs, err := cdilib.GetAllDeviceSpecs()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create device CDI specs: %v", err)
+	}
+
+	commonEdits, err := cdilib.GetCommonEdits()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create edits common for entities: %v", err)
+	}
+
+	spec := &specs.Spec{
+		Version:        cdiSpec.CurrentVersion,
+		Kind:           fmt.Sprintf("%s/%s", cdiID.Vendor, cdiID.Class),
+		Devices:        deviceSpecs,
+		ContainerEdits: *commonEdits.ContainerEdits,
+	}
+
+	logger.Debugf("Generated CDI specification:")
+	logger.Debugf(logger.Pretty(spec))
```

Comment:
Is this a large string? Would it be better and more conventional to write this file out to the inst.LogPath() like we do for other things, like liblxc, proxy device etc?

Reply:
It is a large string indeed (~ 1000 lines).. I agree, we should output it to `inst.LogPath()`

Reply:
fixed